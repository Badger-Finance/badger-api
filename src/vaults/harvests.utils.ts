import { greaterThanOrEqualTo } from '@aws/dynamodb-expressions';
import {
  formatBalance,
  gqlGenT,
  keyBy,
  Vault__factory,
  VaultHarvestData,
  YieldEvent,
  YieldType,
} from '@badger-dao/sdk';
import { BadgerTreeDistribution_OrderBy, SettHarvest_OrderBy } from '@badger-dao/sdk/lib/graphql/generated/badger';
import { BigNumber, ethers } from 'ethers';

import { getDataMapper, getVaultEntityId } from '../aws/dynamodb.utils';
import { VaultDefinitionModel } from '../aws/models/vault-definition.model';
import { VaultYieldEvent } from '../aws/models/vault-yield-event.model';
import { Chain } from '../chains/config/chain.config';
import { OrderDirection } from '../graphql/generated/balancer';
import { queryPriceAtTimestamp } from '../prices/prices.utils';
import { getFullToken } from '../tokens/tokens.utils';
import { getInfuelnceVaultYieldBalance, isInfluenceVault } from './influence.utils';
import { VaultYieldItem } from './interfaces/vault-yield-item.interface';
import { VAULT_TWAY_DURATION } from './vaults.config';
import { getStrategyInfo } from './vaults.utils';
import { calculateYield } from './yields.utils';

// this allows five chunks of 10k blocks per index, approx 2 week
export const HARVEST_SCAN_BLOCK_INCREMENT = 100_000;

function grabYieldGraphTx(graphTxId?: string): string {
  if (!graphTxId) {
    return '';
  }

  if (!graphTxId.includes('-')) {
    return graphTxId;
  }

  return graphTxId.split('-')[0];
}

/**
 * Modify subgraph response to match on chain event data allowing it to fit into our estimation functions.
 * @param vault vault requesting graph data transformation
 * @param harvests harvests data retrieved from the graph
 * @param distributions distribution data retrieved from the graph
 * @returns vault harvest data in the same form as delivered via event logs
 */
function constructGraphVaultData(
  vault: VaultDefinitionModel,
  harvests: gqlGenT.SettHarvestsQuery['settHarvests'],
  distributions: gqlGenT.BadgerTreeDistributionsQuery['badgerTreeDistributions'],
): VaultHarvestData[] {
  const harvestsByTimestamp = keyBy(harvests, (harvest) => harvest.timestamp);
  const treeDistributionsByTimestamp = keyBy(distributions, (distribution) => distribution.timestamp);
  const timestamps = Array.from(
    new Set([...harvestsByTimestamp.keys(), ...treeDistributionsByTimestamp.keys()]).values(),
  );
  return timestamps.map((t) => {
    const timestamp = Number(t);
    const currentHarvests = harvestsByTimestamp.get(timestamp) ?? [];
    const currentDistributions = treeDistributionsByTimestamp.get(timestamp) ?? [];
    return {
      timestamp,
      harvests: currentHarvests.map((h) => ({
        tx: grabYieldGraphTx(h.id),
        timestamp,
        block: Number(h.blockNumber),
        token: vault.depositToken,
        amount: h.amount,
      })),
      treeDistributions: currentDistributions.map((d) => {
        const tokenAddress = d.token.id.startsWith('0x0x') ? d.token.id.slice(2) : d.token.id;
        return {
          tx: grabYieldGraphTx(d.id),
          timestamp,
          block: Number(d.blockNumber),
          token: ethers.utils.getAddress(tokenAddress),
          amount: d.amount,
        };
      }),
    };
  });
}

/**
 *
 * @param data
 * @returns
 */
function constructYieldItems(data: VaultHarvestData[]): VaultYieldItem[] {
  const recentHarvests = data.sort((a, b) => b.timestamp - a.timestamp);
  const allHarvests = recentHarvests.flatMap((h) => h.harvests.map((h) => ({ ...h, type: YieldType.Harvest })));
  const allDistributions = recentHarvests.flatMap((h) =>
    h.treeDistributions.map((d) => ({ ...d, type: YieldType.TreeDistribution })),
  );
  return allHarvests
    .concat(allDistributions)
    .filter((e) => BigNumber.from(e.amount).gt(ethers.constants.Zero))
    .sort((a, b) => b.timestamp - a.timestamp);
}

/**
 *
 * @param chain
 * @param vault
 * @param block
 * @returns
 */
async function getVaultBalance(chain: Chain, vault: VaultDefinitionModel, block: number): Promise<number> {
  const sdk = await chain.getSdk();

  let balance = 0;
  if (isInfluenceVault(vault.address)) {
    balance = await getInfuelnceVaultYieldBalance(chain, vault, block);
  } else {
    const vaultContract = Vault__factory.connect(vault.address, sdk.provider);
    const totalSupply = await vaultContract.totalSupply({ blockTag: block });
    balance = formatBalance(totalSupply);
  }

  return balance;
}

/**
 * Evaluate yield events to derivate base apr, and apy as well as apr of any emitted tokens.
 * @param chain network the vault is deployed on
 * @param vault vault requested yield event evaluation
 * @param yieldEvents events sourced from either on chain or the graph
 * @returns yield summary providing the base information for construction of yield sources
 */
async function evaluateYieldItems(
  chain: Chain,
  vault: VaultDefinitionModel,
  yieldItems: VaultYieldItem[],
): Promise<YieldEvent[]> {
  const yieldEvents: YieldEvent[] = [];
  const tokenEmissionAprs = new Map<string, number>();

  for (const item of yieldItems) {
    const block = item.block;
    const token = await getFullToken(chain, item.token);
    const amount = formatBalance(item.amount, token.decimals);
    const { price } = await queryPriceAtTimestamp(token.address, item.timestamp * 1000);

    const tokenEarned = price * amount;

    const balance = await getVaultBalance(chain, vault, item.block);
    const { price: vaultPrice } = await queryPriceAtTimestamp(vault.address, item.timestamp * 1000);
    const vaultPrincipal = vaultPrice * balance;
    const strategyInfo = await getStrategyInfo(chain, vault, { blockTag: item.block });
    const performanceScalar = 1 / (1 - strategyInfo.performanceFee / 10_000);

    const eventApr = calculateYield(vaultPrincipal, tokenEarned, VAULT_TWAY_DURATION);
    const yieldEvent: YieldEvent = {
      block,
      timestamp: item.timestamp * 1000,
      amount,
      token: token.address,
      type: item.type,
      value: vaultPrincipal,
      balance,
      earned: tokenEarned,
      apr: eventApr,
      grossApr: eventApr * performanceScalar,
      tx: item.tx,
    };
    yieldEvents.push(yieldEvent);

    if (item.type === YieldType.TreeDistribution) {
      const entry = tokenEmissionAprs.get(token.address) ?? 0;
      tokenEmissionAprs.set(token.address, entry + eventApr);
    }
  }

  return yieldEvents;
}

/**
 *
 * @param chain
 * @param vault
 * @param cutoff
 * @returns
 */
async function loadGraphYieldData(
  chain: Chain,
  vault: VaultDefinitionModel,
  cutoff: number,
): Promise<VaultHarvestData[]> {
  const { graph } = await chain.getSdk();
  const { address } = vault;
  console.log(`[${vault.name}]: Load Graph Yield Data`);
  const [vaultHarvests, treeDistributions] = await Promise.all([
    graph.loadSettHarvests({
      first: 25,
      where: {
        sett: address.toLowerCase(),
        timestamp_gt: cutoff,
      },
      orderBy: SettHarvest_OrderBy.Timestamp,
      orderDirection: OrderDirection.Asc,
    }),
    graph.loadBadgerTreeDistributions({
      first: 25,
      where: {
        sett: address.toLowerCase(),
        timestamp_gt: cutoff,
      },
      orderBy: BadgerTreeDistribution_OrderBy.Timestamp,
      orderDirection: OrderDirection.Asc,
    }),
  ]);
  const { settHarvests } = vaultHarvests;
  const { badgerTreeDistributions } = treeDistributions;
  return constructGraphVaultData(vault, settHarvests, badgerTreeDistributions);
}

/**
 *
 * @param chain
 * @param vault
 * @param lastHarvestBlock
 * @param cutoff
 * @returns
 */
async function loadEventYieldData(
  chain: Chain,
  vault: VaultDefinitionModel,
  lastHarvestBlock: number,
  cutoff: number,
): Promise<VaultHarvestData[]> {
  const { vaults } = await chain.getSdk();
  const { address, version } = vault;
  console.log(`[${vault.name}]: Load Event Yield Data`);
  try {
    const { data } = await vaults.listHarvests({
      address,
      timestamp_gte: cutoff,
      version,
      startBlock: lastHarvestBlock,
      endBlock: lastHarvestBlock + HARVEST_SCAN_BLOCK_INCREMENT,
    });
    return data;
  } catch {
    return loadGraphYieldData(chain, vault, cutoff);
  }
}

/**
 *
 * @param chain
 * @param vault
 * @param lastHarvestBlock
 * @returns
 */
export async function loadYieldEvents(
  chain: Chain,
  vault: VaultDefinitionModel,
  lastHarvestBlock: number,
): Promise<YieldEvent[]> {
  const { provider } = await chain.getSdk();
  const block = await provider.getBlock(lastHarvestBlock);
  const cutoff = block.timestamp;

  let data: VaultHarvestData[] = [];

  try {
    data = await loadGraphYieldData(chain, vault, cutoff);
  } catch (err) {
    if (isInfluenceVault(vault.address)) {
      throw err;
    } else {
      data = await loadEventYieldData(chain, vault, lastHarvestBlock, cutoff);
    }
  }

  const yieldItems = constructYieldItems(data);
  return evaluateYieldItems(chain, vault, yieldItems);
}

/**
 *
 * @param chain
 * @param vault
 * @returns
 */
export async function queryVaultYieldEvents(chain: Chain, vault: VaultDefinitionModel): Promise<VaultYieldEvent[]> {
  const mapper = getDataMapper();
  const chainAddress = getVaultEntityId(chain, vault);
  const cutoff = Date.now() - VAULT_TWAY_DURATION;

  const yieldEvents = [];
  for await (const yieldEvent of mapper.query(
    VaultYieldEvent,
    { chainAddress, timestamp: greaterThanOrEqualTo(cutoff) },
    { indexName: 'IndexYieldDataOnAddress', scanIndexForward: false },
  )) {
    yieldEvents.push(yieldEvent);
  }
  return yieldEvents;
}

/**
 *
 * @param chain
 * @param vault
 * @returns
 */
export async function queryVaultHistoricYieldEvents(
  chain: Chain,
  vault: VaultDefinitionModel,
): Promise<VaultYieldEvent[]> {
  // TODO: construct a new paginated 'all harvests' table
  // allow for page queries, timestamp based page queries
  // all pages will be the same size, new pages are only added after x entries to a given page
  // we can follow similar to the charting pattern with this data
  const mapper = getDataMapper();
  const chainAddress = getVaultEntityId(chain, vault);

  const yieldEvents = [];
  for await (const yieldEvent of mapper.query(
    VaultYieldEvent,
    { chainAddress },
    { indexName: 'IndexYieldDataOnAddress', scanIndexForward: false },
  )) {
    yieldEvents.push(yieldEvent);
  }
  return yieldEvents;
}
