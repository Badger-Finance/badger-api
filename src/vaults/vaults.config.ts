import { ONE_DAY_SECONDS } from '@badger-dao/sdk';

export const VAULT_TWAY_PERIOD = 15;
export const VAULT_TWAY_DURATION = VAULT_TWAY_PERIOD * ONE_DAY_SECONDS;
export const VAULT_SOURCE = 'Vault Compounding';
