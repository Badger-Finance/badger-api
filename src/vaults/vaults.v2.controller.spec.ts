import { DataMapper, QueryIterator, StringToAnyObjectMap } from '@aws/dynamodb-data-mapper';
import BadgerSDK, { ONE_DAY_MS, TokenValue, VaultDTO } from '@badger-dao/sdk';
import { PlatformTest } from '@tsed/common';
import { BadRequest } from '@tsed/exceptions';
import createMockInstance from 'jest-create-mock-instance';
import SuperTest from 'supertest';

import { VaultDefinitionModel } from '../aws/models/vault-definition.model';
import { VaultPendingHarvestData } from '../aws/models/vault-pending-harvest.model';
import { YieldSource } from '../aws/models/yield-source.model';
import { Chain } from '../chains/config/chain.config';
import { TOKENS } from '../config/tokens.config';
import { SourceType } from '../rewards/enums/source-type.enum';
import { Server } from '../Server';
import { mockChainVaults } from '../test/tests.utils';
import { fullTokenMockMap } from '../tokens/mocks/full-token.mock';
import * as tokensUtils from '../tokens/tokens.utils';
import { mockBalance } from '../tokens/tokens.utils';
import { vaultsHarvestsMapMock } from './mocks/vaults-harvests-map.mock';
import * as vaultsUtils from './vaults.utils';

const TEST_VAULT = TOKENS.BCRV_SBTC;

export function setupDdbHarvests() {
  mockChainVaults();
  jest.spyOn(BadgerSDK.prototype, 'ready').mockImplementation();

  /* eslint-disable @typescript-eslint/ban-ts-comment */
  // @ts-ignore
  jest.spyOn(DataMapper.prototype, 'query').mockImplementation((_model, _condition) => {
    // @ts-ignore
    const qi: QueryIterator<StringToAnyObjectMap> = createMockInstance(QueryIterator);
    // @ts-ignore
    qi[Symbol.iterator] = jest.fn(() => {
      return (vaultsHarvestsMapMock[_condition.vault] || []).values();
    });

    return qi;
  });
  /* eslint-enable @typescript-eslint/ban-ts-comment */
}

export function setupTestVault() {
  mockChainVaults();
  jest.spyOn(tokensUtils, 'getFullToken').mockImplementation(async (_, tokenAddr) => {
    return fullTokenMockMap[tokenAddr] || fullTokenMockMap[TOKENS.BADGER];
  });
  const baseTime = 1656606946;
  jest.spyOn(Date, 'now').mockImplementation(() => baseTime * 1000 + ONE_DAY_MS * 14);
  jest.spyOn(vaultsUtils, 'getVaultPendingHarvest').mockImplementation(
    async (vaultDefinition: VaultDefinitionModel): Promise<VaultPendingHarvestData> => ({
      vault: vaultDefinition.address,
      yieldTokens: [mockBalance(fullTokenMockMap[TOKENS.CVX], 10)],
      harvestTokens: [mockBalance(fullTokenMockMap[TOKENS.CVX], 10)],
      lastHarvestedAt: baseTime,
      lastMeasuredAt: baseTime,
      previousYieldTokens: [mockBalance(fullTokenMockMap[TOKENS.CVX], 10)],
      previousHarvestTokens: [mockBalance(fullTokenMockMap[TOKENS.CVX], 10)],
      duration: 60000,
      lastReportedAt: 0,
    }),
  );
  jest
    .spyOn(vaultsUtils, 'getCachedVault')
    .mockImplementation(async (chain: Chain, vaultDefinition: VaultDefinitionModel): Promise<VaultDTO> => {
      const vault = await vaultsUtils.defaultVault(chain, vaultDefinition);
      vault.value = parseInt(vaultDefinition.address.slice(0, 7), 16);
      vault.balance = 10;
      return vault;
    });
  jest
    .spyOn(vaultsUtils, 'getVaultYieldSources')
    .mockImplementation(async (vault: VaultDefinitionModel): Promise<YieldSource[]> => {
      const performance = parseInt(vault.address.slice(0, 5), 16) / 100;
      const underlying = vaultsUtils.createYieldSource(
        vault.address,
        SourceType.Compound,
        vaultsUtils.VAULT_SOURCE,
        performance,
      );
      const badger = vaultsUtils.createYieldSource(vault.address, SourceType.Emission, 'Badger Rewards', performance);
      const fees = vaultsUtils.createYieldSource(vault.address, SourceType.TradeFee, 'Curve LP Fees', performance);
      return [underlying, badger, fees];
    });
  jest
    .spyOn(tokensUtils, 'getVaultTokens')
    .mockImplementation(async (_chain: Chain, vault: VaultDTO, _currency?: string): Promise<TokenValue[]> => {
      const token = fullTokenMockMap[vault.underlyingToken] || fullTokenMockMap[TOKENS.BADGER];
      if (token.lpToken) {
        const bal0 = parseInt(token.address.slice(0, 4), 16);
        const bal1 = parseInt(token.address.slice(0, 6), 16);
        return [mockBalance(token, bal0), mockBalance(token, bal1)];
      }
      return [mockBalance(token, parseInt(token.address.slice(0, 4), 16))];
    });
}

describe('VaultsController', () => {
  let request: SuperTest.SuperTest<SuperTest.Test>;

  beforeAll(PlatformTest.bootstrap(Server));
  beforeAll(async () => {
    request = SuperTest(PlatformTest.callback());
  });
  afterAll(PlatformTest.reset);

  describe('GET /v2/vaults', () => {
    describe('with no specified chain', () => {
      it('returns eth vaults', async (done: jest.DoneCallback) => {
        setupTestVault();
        const { body } = await request.get('/v2/vaults').expect(200);
        expect(body).toMatchSnapshot();
        done();
      });
    });

    describe('with a specified chain', () => {
      it('returns the vaults for eth', async (done: jest.DoneCallback) => {
        setupTestVault();
        const { body } = await request.get('/v2/vaults?chain=ethereum').expect(200);
        expect(body).toMatchSnapshot();
        done();
      });

      it('returns the vaults for bsc', async (done: jest.DoneCallback) => {
        setupTestVault();
        const { body } = await request.get('/v2/vaults?chain=binancesmartchain').expect(200);
        expect(body).toMatchSnapshot();
        done();
      });
    });

    describe('with an invalid specified chain', () => {
      it('returns a 400', async (done: jest.DoneCallback) => {
        const { body } = await request.get('/v2/vaults?chain=invalid').expect(BadRequest.STATUS);
        expect(body).toMatchSnapshot();
        done();
      });
    });
  });

  describe('GET /v2/harvests', () => {
    beforeEach(setupDdbHarvests);

    describe('success cases', () => {
      it('Return extended harvest for chain vaults', async (done: jest.DoneCallback) => {
        const { body } = await request.get('/v2/vaults/harvests').expect(200);
        expect(body).toMatchSnapshot();
        done();
      });
    });
    describe('error cases', () => {
      it('returns a 400 for invalide chain', async (done: jest.DoneCallback) => {
        const { body } = await request.get('/v2/vaults/harvests?chain=invalid').expect(BadRequest.STATUS);
        expect(body).toMatchSnapshot();
        done();
      });
    });
  });

  describe('GET /v2/harvests/:vault', () => {
    beforeEach(setupDdbHarvests);

    describe('success cases', () => {
      it('Return extended harvests for chain vault by addr', async (done: jest.DoneCallback) => {
        const { body } = await request.get(`/v2/vaults/harvests/${TEST_VAULT}`).expect(200);
        expect(body).toMatchSnapshot();
        done();
      });
    });
    describe('error cases', () => {
      it('returns a 400 for invalide chain', async (done: jest.DoneCallback) => {
        const { body } = await request.get(`/v2/vaults/harvests/${TEST_VAULT}?chain=invalid`).expect(BadRequest.STATUS);
        expect(body).toMatchSnapshot();
        done();
      });
    });
  });
});
