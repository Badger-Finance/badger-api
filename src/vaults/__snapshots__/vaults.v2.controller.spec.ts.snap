// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VaultsController GET /v2/harvests error cases returns a 400 for invalide chain 1`] = `
Object {
  "message": "invalid is not a supported chain",
}
`;

exports[`VaultsController GET /v2/harvests success cases Return extended harvest for chain vaults 1`] = `
Object {
  "0x19D97D8fA813EE2f51aD4B4e04EA08bAf4DFfC28": Array [],
}
`;

exports[`VaultsController GET /v2/harvests/:vault error cases returns a 400 for invalide chain 1`] = `
Object {
  "message": "invalid is not a supported chain",
}
`;

exports[`VaultsController GET /v2/harvests/:vault success cases Return extended harvests for chain vault by addr 1`] = `Array []`;

exports[`VaultsController GET /v2/vaults with a specified chain returns the vaults for bsc 1`] = `
Array [
  Object {
    "apr": 8.26,
    "apy": 12.39,
    "asset": "BADGER",
    "available": 0,
    "balance": 10,
    "behavior": "Compounder",
    "boost": Object {
      "enabled": false,
      "weight": 0,
    },
    "bouncer": "none",
    "lastHarvest": 1656606946,
    "name": "Badger",
    "pricePerFullShare": 1,
    "protocol": "Badger",
    "sources": Array [
      Object {
        "apr": 4.13,
        "boostable": false,
        "maxApr": 4.13,
        "minApr": 4.13,
        "name": "Badger Rewards",
      },
      Object {
        "apr": 4.13,
        "boostable": false,
        "maxApr": 4.13,
        "minApr": 4.13,
        "name": "Curve LP Fees",
      },
    ],
    "sourcesApy": Array [
      Object {
        "apr": 4.13,
        "boostable": false,
        "maxApr": 4.13,
        "minApr": 4.13,
        "name": "Vault Compounding",
      },
      Object {
        "apr": 4.13,
        "boostable": false,
        "maxApr": 4.13,
        "minApr": 4.13,
        "name": "Badger Rewards",
      },
      Object {
        "apr": 4.13,
        "boostable": false,
        "maxApr": 4.13,
        "minApr": 4.13,
        "name": "Curve LP Fees",
      },
    ],
    "state": "open",
    "strategy": Object {
      "address": "0x0000000000000000000000000000000000000000",
      "aumFee": 0,
      "performanceFee": 0,
      "strategistFee": 0,
      "withdrawFee": 0,
    },
    "tokens": Array [],
    "type": "native",
    "underlyingToken": "0x3472A5A71965499acd81997a54BBA8D852C6E53d",
    "value": 105879,
    "vaultAsset": "bBADGER",
    "vaultToken": "0x19D97D8fA813EE2f51aD4B4e04EA08bAf4DFfC28",
    "version": "v1",
    "yieldProjection": Object {
      "harvestApr": 0.22498409152323756,
      "harvestPeriodApr": 0,
      "harvestPeriodApy": 0,
      "harvestPeriodSources": Array [
        Object {
          "address": "0x4e3FBD56CD56c3e72c1403e103b45Db9da5B9D2B",
          "apr": 0,
          "balance": 0,
          "decimals": 18,
          "name": "Convex Token",
          "symbol": "CVX",
          "value": 0,
        },
      ],
      "harvestPeriodSourcesApy": Array [
        Object {
          "address": "0x4e3FBD56CD56c3e72c1403e103b45Db9da5B9D2B",
          "apr": 0,
          "balance": 0,
          "decimals": 18,
          "name": "Convex Token",
          "symbol": "CVX",
          "value": 0,
        },
      ],
      "harvestTokens": Array [
        Object {
          "address": "0x4e3FBD56CD56c3e72c1403e103b45Db9da5B9D2B",
          "apr": 0.22498409152323756,
          "balance": 10,
          "decimals": 18,
          "name": "Convex Token",
          "symbol": "CVX",
          "value": 12510,
        },
      ],
      "harvestValue": 12510,
      "nonHarvestApr": 8.26,
      "nonHarvestApy": 8.26,
      "nonHarvestSources": Array [
        Object {
          "apr": 4.13,
          "boostable": false,
          "maxApr": 4.13,
          "minApr": 4.13,
          "name": "Badger Rewards",
        },
        Object {
          "apr": 4.13,
          "boostable": false,
          "maxApr": 4.13,
          "minApr": 4.13,
          "name": "Curve LP Fees",
        },
      ],
      "nonHarvestSourcesApy": Array [
        Object {
          "apr": 4.13,
          "boostable": false,
          "maxApr": 4.13,
          "minApr": 4.13,
          "name": "Badger Rewards",
        },
        Object {
          "apr": 4.13,
          "boostable": false,
          "maxApr": 4.13,
          "minApr": 4.13,
          "name": "Curve LP Fees",
        },
      ],
      "yieldApr": 0.22498409152323756,
      "yieldPeriodApr": 0,
      "yieldPeriodSources": Array [
        Object {
          "address": "0x4e3FBD56CD56c3e72c1403e103b45Db9da5B9D2B",
          "apr": 0,
          "balance": 0,
          "decimals": 18,
          "name": "Convex Token",
          "symbol": "CVX",
          "value": 0,
        },
      ],
      "yieldTokens": Array [
        Object {
          "address": "0x4e3FBD56CD56c3e72c1403e103b45Db9da5B9D2B",
          "apr": 0.22498409152323756,
          "balance": 10,
          "decimals": 18,
          "name": "Convex Token",
          "symbol": "CVX",
          "value": 12510,
        },
      ],
      "yieldValue": 12510,
    },
  },
]
`;

exports[`VaultsController GET /v2/vaults with a specified chain returns the vaults for eth 1`] = `
Array [
  Object {
    "apr": 8.26,
    "apy": 12.39,
    "asset": "BADGER",
    "available": 0,
    "balance": 10,
    "behavior": "Compounder",
    "boost": Object {
      "enabled": false,
      "weight": 0,
    },
    "bouncer": "none",
    "lastHarvest": 1656606946,
    "name": "Badger",
    "pricePerFullShare": 1,
    "protocol": "Badger",
    "sources": Array [
      Object {
        "apr": 4.13,
        "boostable": false,
        "maxApr": 4.13,
        "minApr": 4.13,
        "name": "Badger Rewards",
      },
      Object {
        "apr": 4.13,
        "boostable": false,
        "maxApr": 4.13,
        "minApr": 4.13,
        "name": "Curve LP Fees",
      },
    ],
    "sourcesApy": Array [
      Object {
        "apr": 4.13,
        "boostable": false,
        "maxApr": 4.13,
        "minApr": 4.13,
        "name": "Vault Compounding",
      },
      Object {
        "apr": 4.13,
        "boostable": false,
        "maxApr": 4.13,
        "minApr": 4.13,
        "name": "Badger Rewards",
      },
      Object {
        "apr": 4.13,
        "boostable": false,
        "maxApr": 4.13,
        "minApr": 4.13,
        "name": "Curve LP Fees",
      },
    ],
    "state": "open",
    "strategy": Object {
      "address": "0x0000000000000000000000000000000000000000",
      "aumFee": 0,
      "performanceFee": 0,
      "strategistFee": 0,
      "withdrawFee": 0,
    },
    "tokens": Array [],
    "type": "native",
    "underlyingToken": "0x3472A5A71965499acd81997a54BBA8D852C6E53d",
    "value": 105879,
    "vaultAsset": "bBADGER",
    "vaultToken": "0x19D97D8fA813EE2f51aD4B4e04EA08bAf4DFfC28",
    "version": "v1",
    "yieldProjection": Object {
      "harvestApr": 0.22498409152323756,
      "harvestPeriodApr": 0,
      "harvestPeriodApy": 0,
      "harvestPeriodSources": Array [
        Object {
          "address": "0x4e3FBD56CD56c3e72c1403e103b45Db9da5B9D2B",
          "apr": 0,
          "balance": 0,
          "decimals": 18,
          "name": "Convex Token",
          "symbol": "CVX",
          "value": 0,
        },
      ],
      "harvestPeriodSourcesApy": Array [
        Object {
          "address": "0x4e3FBD56CD56c3e72c1403e103b45Db9da5B9D2B",
          "apr": 0,
          "balance": 0,
          "decimals": 18,
          "name": "Convex Token",
          "symbol": "CVX",
          "value": 0,
        },
      ],
      "harvestTokens": Array [
        Object {
          "address": "0x4e3FBD56CD56c3e72c1403e103b45Db9da5B9D2B",
          "apr": 0.22498409152323756,
          "balance": 10,
          "decimals": 18,
          "name": "Convex Token",
          "symbol": "CVX",
          "value": 12510,
        },
      ],
      "harvestValue": 12510,
      "nonHarvestApr": 8.26,
      "nonHarvestApy": 8.26,
      "nonHarvestSources": Array [
        Object {
          "apr": 4.13,
          "boostable": false,
          "maxApr": 4.13,
          "minApr": 4.13,
          "name": "Badger Rewards",
        },
        Object {
          "apr": 4.13,
          "boostable": false,
          "maxApr": 4.13,
          "minApr": 4.13,
          "name": "Curve LP Fees",
        },
      ],
      "nonHarvestSourcesApy": Array [
        Object {
          "apr": 4.13,
          "boostable": false,
          "maxApr": 4.13,
          "minApr": 4.13,
          "name": "Badger Rewards",
        },
        Object {
          "apr": 4.13,
          "boostable": false,
          "maxApr": 4.13,
          "minApr": 4.13,
          "name": "Curve LP Fees",
        },
      ],
      "yieldApr": 0.22498409152323756,
      "yieldPeriodApr": 0,
      "yieldPeriodSources": Array [
        Object {
          "address": "0x4e3FBD56CD56c3e72c1403e103b45Db9da5B9D2B",
          "apr": 0,
          "balance": 0,
          "decimals": 18,
          "name": "Convex Token",
          "symbol": "CVX",
          "value": 0,
        },
      ],
      "yieldTokens": Array [
        Object {
          "address": "0x4e3FBD56CD56c3e72c1403e103b45Db9da5B9D2B",
          "apr": 0.22498409152323756,
          "balance": 10,
          "decimals": 18,
          "name": "Convex Token",
          "symbol": "CVX",
          "value": 12510,
        },
      ],
      "yieldValue": 12510,
    },
  },
]
`;

exports[`VaultsController GET /v2/vaults with an invalid specified chain returns a 400 1`] = `
Object {
  "message": "invalid is not a supported chain",
}
`;

exports[`VaultsController GET /v2/vaults with no specified chain returns eth vaults 1`] = `
Array [
  Object {
    "apr": 8.26,
    "apy": 12.39,
    "asset": "BADGER",
    "available": 0,
    "balance": 10,
    "behavior": "Compounder",
    "boost": Object {
      "enabled": false,
      "weight": 0,
    },
    "bouncer": "none",
    "lastHarvest": 1656606946,
    "name": "Badger",
    "pricePerFullShare": 1,
    "protocol": "Badger",
    "sources": Array [
      Object {
        "apr": 4.13,
        "boostable": false,
        "maxApr": 4.13,
        "minApr": 4.13,
        "name": "Badger Rewards",
      },
      Object {
        "apr": 4.13,
        "boostable": false,
        "maxApr": 4.13,
        "minApr": 4.13,
        "name": "Curve LP Fees",
      },
    ],
    "sourcesApy": Array [
      Object {
        "apr": 4.13,
        "boostable": false,
        "maxApr": 4.13,
        "minApr": 4.13,
        "name": "Vault Compounding",
      },
      Object {
        "apr": 4.13,
        "boostable": false,
        "maxApr": 4.13,
        "minApr": 4.13,
        "name": "Badger Rewards",
      },
      Object {
        "apr": 4.13,
        "boostable": false,
        "maxApr": 4.13,
        "minApr": 4.13,
        "name": "Curve LP Fees",
      },
    ],
    "state": "open",
    "strategy": Object {
      "address": "0x0000000000000000000000000000000000000000",
      "aumFee": 0,
      "performanceFee": 0,
      "strategistFee": 0,
      "withdrawFee": 0,
    },
    "tokens": Array [],
    "type": "native",
    "underlyingToken": "0x3472A5A71965499acd81997a54BBA8D852C6E53d",
    "value": 105879,
    "vaultAsset": "bBADGER",
    "vaultToken": "0x19D97D8fA813EE2f51aD4B4e04EA08bAf4DFfC28",
    "version": "v1",
    "yieldProjection": Object {
      "harvestApr": 0.22498409152323756,
      "harvestPeriodApr": 0,
      "harvestPeriodApy": 0,
      "harvestPeriodSources": Array [
        Object {
          "address": "0x4e3FBD56CD56c3e72c1403e103b45Db9da5B9D2B",
          "apr": 0,
          "balance": 0,
          "decimals": 18,
          "name": "Convex Token",
          "symbol": "CVX",
          "value": 0,
        },
      ],
      "harvestPeriodSourcesApy": Array [
        Object {
          "address": "0x4e3FBD56CD56c3e72c1403e103b45Db9da5B9D2B",
          "apr": 0,
          "balance": 0,
          "decimals": 18,
          "name": "Convex Token",
          "symbol": "CVX",
          "value": 0,
        },
      ],
      "harvestTokens": Array [
        Object {
          "address": "0x4e3FBD56CD56c3e72c1403e103b45Db9da5B9D2B",
          "apr": 0.22498409152323756,
          "balance": 10,
          "decimals": 18,
          "name": "Convex Token",
          "symbol": "CVX",
          "value": 12510,
        },
      ],
      "harvestValue": 12510,
      "nonHarvestApr": 8.26,
      "nonHarvestApy": 8.26,
      "nonHarvestSources": Array [
        Object {
          "apr": 4.13,
          "boostable": false,
          "maxApr": 4.13,
          "minApr": 4.13,
          "name": "Badger Rewards",
        },
        Object {
          "apr": 4.13,
          "boostable": false,
          "maxApr": 4.13,
          "minApr": 4.13,
          "name": "Curve LP Fees",
        },
      ],
      "nonHarvestSourcesApy": Array [
        Object {
          "apr": 4.13,
          "boostable": false,
          "maxApr": 4.13,
          "minApr": 4.13,
          "name": "Badger Rewards",
        },
        Object {
          "apr": 4.13,
          "boostable": false,
          "maxApr": 4.13,
          "minApr": 4.13,
          "name": "Curve LP Fees",
        },
      ],
      "yieldApr": 0.22498409152323756,
      "yieldPeriodApr": 0,
      "yieldPeriodSources": Array [
        Object {
          "address": "0x4e3FBD56CD56c3e72c1403e103b45Db9da5B9D2B",
          "apr": 0,
          "balance": 0,
          "decimals": 18,
          "name": "Convex Token",
          "symbol": "CVX",
          "value": 0,
        },
      ],
      "yieldTokens": Array [
        Object {
          "address": "0x4e3FBD56CD56c3e72c1403e103b45Db9da5B9D2B",
          "apr": 0.22498409152323756,
          "balance": 10,
          "decimals": 18,
          "name": "Convex Token",
          "symbol": "CVX",
          "value": 12510,
        },
      ],
      "yieldValue": 12510,
    },
  },
]
`;
